src/py/flwr/client/middleware/secure_aggregation/secaggplus_middleware.py

```diff
@@ -112,70 +113,115 @@ class SecAggPlusState:
     ss2_dict: Dict[int, bytes] = field(default_factory=dict)
     public_keys_dict: Dict[int, Tuple[bytes, bytes]] = field(default_factory=dict)
 
-    client: Optional[Union[Client, NumPyClient]] = None
-
-
-class SecAggPlusHandler(SecureAggregationHandler):
-    """Message handler for the SecAgg+ protocol."""
-
-    _shared_state = SecAggPlusState()
-    _current_stage = STAGE_UNMASK
-
-    def handle_secure_aggregation(
-        self, named_values: Dict[str, Value]
-    ) -> Dict[str, Value]:
-        """Handle incoming message and return results, following the SecAgg+ protocol.
-
-        Parameters
-        ----------
-        named_values : Dict[str, Value]
-            The named values retrieved from the SecureAggregation sub-message
-            of Task message in the server's TaskIns.
+    def __init__(self, **kwargs: ConfigsRecordValues) -> None:
+        for k, v in kwargs.items():
+            if k.endswith(":V"):
+                continue
+            new_v: Any = v
+            if k.endswith(":K"):
+                k = k[:-2]
+                keys = cast(List[int], v)
+                values = cast(List[bytes], kwargs[f"{k}:V"])
+                if len(values) > len(keys):
+                    updated_values = [
+                        tuple(values[i : i + 2]) for i in range(0, len(values), 2)
+                    ]
+                    new_v = dict(zip(keys, updated_values))
+                else:
+                    new_v = dict(zip(keys, values))
+            self.__setattr__(k, new_v)
+
+    def to_dict(self) -> Dict[str, ConfigsRecordValues]:
+        """Convert the state to a dictionary."""
+        ret = vars(self)
+        for k in list(ret.keys()):
+            if isinstance(ret[k], dict):
+                # Replace dict with two lists
+                v = cast(Dict[str, Any], ret.pop(k))
+                ret[f"{k}:K"] = list(v.keys())
+                if k == "public_keys_dict":
+                    v_list: List[bytes] = []
+                    for b1_b2 in cast(List[Tuple[bytes, bytes]], v.values()):
+                        v_list.extend(b1_b2)
+                    ret[f"{k}:V"] = v_list
+                else:
+                    ret[f"{k}:V"] = list(v.values())
+        return ret
+
+
+def _get_fit_fn(
+    msg: Message, call_next: FlowerCallable, ctxt: Context
+) -> Callable[[], FitRes]:
+    """Get the fit function."""
+
+    def fit() -> FitRes:
+        out_msg = call_next(msg, ctxt)
+        return compat.recordset_to_fitres(out_msg.message, keep_input=False)
```

Comment:
Middleware layers should only use `RecordSet`, but none of the legacy messages like `FitRes`.

Reply:
This is b/c SA MW is a layer wrapping around the `fit` method. The SA MW needs to pass the message to the inner `fit` method to train the model and obtain the updated parameters. But for now, our `RecordSet` APIs do not allow us to easily and directly read/write parameters, so here it's a bit hard to directly modify the parameters inside `RecordSet`. Plus, the returned message is actually converted from a FitRes, so I converted it back to access the updated parameters.

Reply:
This would mean that our SA middleware is incompatible with all future task handlers that do not return a `RecordSet` that was derived from a `FitRes`.

In the worst case, we can update the parameters in the `RecordSet` by constructing a new `RecordSet` that has all the same fields as the `RecordSet` we get from `call_next`, just with modified parameters.